### DISCUSSION TITLE
CF Peer-to-peer communication

### DISCUSSION THREAD

**Discussion by imreburn** [bitcraze/discussions#790] [Q&A] [answered]
Hello everyone,

We would like to build an indoor testbed of swarm of small drones and consider Crazyflies. One of the key features we need is drone-to-drone communication.

I found that broadcast is provided over Crazyradio using CRTP in the [documentation](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/p2p_api/). And unicast feature is under development. I came up with a few ideas to implement this on our own.

1. Ad-hoc network over Wi-Fi using AI decks
I found that Wi-Fi is enabled using an AI deck and CF can send messages via [CPX protocol](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/cpx/). There is an [example](https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/simple-examples/wifi-streamer/) that CF can be a Wi-Fi access point and stream images to the host. But I am not sure that CF can also _receive_ CPX messages from the host or other CFs. 

2. Workarounds via host
If the network b/w CFs cannot be established, I think at least I can mimic the network through host. I found a [paper](https://arxiv.org/abs/2203.16960) did a similar approach. I think there are two ways to do it.
- via Crazyradio (Bluetooth)
- via Wi-Fi using AI decks (if CF can get messages from the host)

I hope to hear similar experiences or comments on this plan before we make purchase. I would greatly appreciate it if I can hear any insights. I am sorry if it sounds a bit abstract. Thank you!

**Comment by ataffanel** [comment, marked as answer]
Hi,

Both idea would work and make sense. As you noticed, using the ground as a relay has already been used and should work well. Just note that we are not using Bluetooth between Crazyradio and Crazyflie but a custom star-network protocol. Over wifi on the AI deck will require extra code though since this is not something that is currently supported (using the `ESP-Now` protocol might actually make a lot of sense here, since `ESP-Now` will allow direct communication between the esp32s). However using the AI-Deck just for that would be a bit overkill in terms of cost and complexity.

While the current P2P support is broadcast only, it is possible to send unicast packet by just adding a destination address in the data. There is no automatic ack but this can also be added on the software side. For reference, we have successfully used the P2P communication in the demo we show at exhibitions, the [decentralized swarm demo](https://github.com/bitcraze/crazyflie-firmware-experimental/tree/icra-2023/examples/demos/decentralized_swarm). In this demo Crazyflies [sends their state to each-other](https://github.com/bitcraze/crazyflie-firmware-experimental/blob/c873bde6b9fa45118bc8f3d7bd736ce5f3b01e3e/examples/demos/decentralized_swarm/src/common_files/ds_p2p_interface.c#L138) at about 15Hz.

We are planning on working soon on a better P2P communication protocol, as part of the work related to the new Crazyradio2 dongle, but I am not sure when we will be able to get to it. In the mean time there is always the possibility to do it yourself since everything is open-source, we know of some university that have done just that (communication where one of their research interest though). My plan is to implement a more 'proper' `CSMA-CA` by listening to the air before sending packet, and have Crazyradio 2 able to access Crazyflies the same way (currently it is hard to get Crazyradio and P2P working at the same time since none of them wait for the air to be free to send a packet, so they collides a lot).

Using the internal radio would work well as long as your bandwidth/packet rate is good enough for your application (I expect that with a properly implemented P2P protocol, 200-300 packets per seconds for the full network should be feasible). If not, using ESP32 is a good idea either with the AI-Deck, or, if you have electronic knowledge on your side, making a deck with an ESP32 module is quite easy either by making a custom PCB or by soldering an esp32 breakout board to the Crazyflie breakout deck. However, in that case, software will be left to you to write (the CPX firmware running on the AI-deck is most likely portable but I am not sure how easy it would be).

I would be interested to know a bit more about what kind of communication performance you would need. How much data and how often will you need to communicate? Is latency very important?

I hope I could help giving you an idea of the current state and what is possible to do. Do not hesitate if you have further questions.

**Reply by imreburn** [reply]
Hi Arnaud,

Thank you a lot for the insightful comments and corrections! Your comments and demos give us confidence to make our research project work with CFs. It is so good to hear that sending unicast packets without ground relay is also possible. Currently we would like to start with broadcasting positioning data like the demo. And we plan to add some command packets later. Considering cost, complexity and latency, I think using internal radio communication is enough for our purpose where about 8 drones broadcast or communicate with each other.

I wonder where I can get more information about the internal radio, especially about the lower layers. I found it uses 2.4 GHz ISM band. Would it be possible to see analog signal received by nRF51 (CF) or nRF52 (Crazyradio) and how it is translated to the bitstream? I think why not as debugging adapter is provided and firmware is open-source. I think this sounds a bit experimental. I would appreciate any comment or information.

**Reply by ataffanel** [reply]
You cannot have this kind of low level access to the radio analog part. The radio in the nRF5 chips is quite flexible and well documented (the reason we are using nRF chips is that this is one of the rare radio chip that actually has a documentation for the radio, other manufacturer tends to just distribute protocols as binary blob). You can always observer the packets with an SDR but the control you have on the analog signal is quite limited.

You can look at the nRF51 reference manual and nRF52840 product specification to see what kind of access we have to the radio. The most important thing with these chip is to actually read the doc: the radio hardware runs a state-machine that can be automated with Nordic's peripheral interface. The software is setting-up and steering the hardware, the hardware is doing quite a lot automatically. So with this radio, reading the software is not enough to understand everything that happens. This is not hard but very powerful so it is important to read the doc :-).

The main control we have over the radio is high-level stuff like channel/bit-rate, packet format and transmission timing. Everything else (modulation, packet serialization and deserialization, CRC check) is handled by the hardware.

For reference you can look at the [Crazyflie](https://github.com/bitcraze/crazyflie2-nrf-firmware/blob/master/src/esb.c) and [Crazyradio](https://github.com/bitcraze/crazyradio2-firmware/blob/main/src/esb.c) implementation of the current protocol. This [blog post](https://www.bitcraze.io/2020/08/crazyflie-radio-communication/) also explains a bit the current radio protocol implementation from Crazyradio to Crazyflie. Crazyradio 2.0 currently implements this protocol. The Crazyradio code might be easier to read, it is more recent and does less, so it is a good example of what can be done with the radio.

**Reply by imreburn** [reply]
Thank you a lot for comments and links for resources. I will check the reference manuals for nRF chips if I can just watch what is happening on hardware side.
